etcdserver/snapshot_hub.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+type snapshot struct {
+	// storing the raft snapshot. the data in raftsnap is v2 store.
+	raftsnap raftpb.Snapshot
+	// rc is a reader that provides the data of KV snapshot.
+	rc io.ReadCloser
+	// dataSize specifies the size of snapshot data.
```

Comment:
After thinking about it, I still want to let rafthttp decide how to send it, because this part of logic is common if we have raftdb.